subsys/bluetooth/host/ecc.c

```diff
@@ -0,0 +1,202 @@
+}
+
+int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
+{
+	int err;
+
+	/*
+	 * We check for both "LE Read Local P-256 Public Key" and
+	 * "LE Generate DH Key" support here since both commands are needed for
+	 * ECC support. If "LE Generate DH Key" is not supported then there
+	 * is no point in reading local public key.
+	 */
+	if (!BT_CMD_TEST(bt_dev.supported_commands, 34, 1) ||
+	    !BT_CMD_TEST(bt_dev.supported_commands, 34, 2)) {
+		BT_WARN("ECC HCI commands not available");
+		return -ENOTSUP;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_USE_DEBUG_KEYS)) {
+		if (!BT_CMD_TEST(bt_dev.supported_commands, 41, 2)) {
+			BT_WARN("ECC Debug keys HCI command not available");
+		} else {
+			atomic_set_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY);
+			new_cb->func(debug_public_key);
+			return 0;
+		}
+	}
+
+	new_cb->_next = pub_key_cb;
+	pub_key_cb = new_cb;
+
+	if (atomic_test_and_set_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY)) {
+		return 0;
+	}
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_P256_PUBLIC_KEY, NULL, NULL);
+	if (err) {
+		BT_ERR("Sending LE P256 Public Key command failed");
+		atomic_clear_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY);
+		pub_key_cb = NULL;
+		return err;
+	}
+
+	return 0;
+}
+
+const uint8_t *bt_pub_key_get(void)
+{
+	if (IS_ENABLED(CONFIG_BT_USE_DEBUG_KEYS) &&
+	    BT_CMD_TEST(bt_dev.supported_commands, 41, 2)) {
+		return debug_public_key;
+	}
+
+	if (atomic_test_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY)) {
+		return pub_key;
+	}
+
+	return NULL;
+}
+
+static int hci_generate_dhkey_v1(const uint8_t *remote_pk)
+{
+	struct bt_hci_cp_le_generate_dhkey *cp;
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_GENERATE_DHKEY, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	memcpy(cp->key, remote_pk, sizeof(cp->key));
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_GENERATE_DHKEY, buf, NULL);
+}
+
+static int hci_generate_dhkey_v2(const uint8_t *remote_pk, uint8_t key_type)
+{
+	struct bt_hci_cp_le_generate_dhkey_v2 *cp;
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_GENERATE_DHKEY_V2, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	memcpy(cp->key, remote_pk, sizeof(cp->key));
+	cp->key_type = key_type;
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_GENERATE_DHKEY_V2, buf, NULL);
+}
+
+int bt_dh_key_gen(const uint8_t remote_pk[64], bt_dh_key_cb_t cb)
+{
+	int err;
+
+	if (dh_key_cb == cb) {
+		return -EALREADY;
+	}
+
+	if (dh_key_cb || atomic_test_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY)) {
+		return -EBUSY;
+	}
+
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_HAS_PUB_KEY)) {
+		return -EADDRNOTAVAIL;
+	}
+
+	dh_key_cb = cb;
+
+	if (IS_ENABLED(CONFIG_BT_USE_DEBUG_KEYS) &&
+	    BT_CMD_TEST(bt_dev.supported_commands, 41, 2)) {
+		err = hci_generate_dhkey_v2(remote_pk,
+					    BT_HCI_LE_KEY_TYPE_DEBUG);
+	} else {
+		err = hci_generate_dhkey_v1(remote_pk);
```

Comment:
warn about inability to use debug keys?

Reply:
There is a warning on this when generating the public key, I thought that was enough, so I skipped it here.
But I can add it here as well if that is preferred.